File: Algorithms/Data Structure dependent/Stacks and Queue/minRemoveToMakeValid.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        int closeCount = 0;
        for(auto &c : s){
            if(c == ')') closeCount++;
        }
        /* track which one are open*/
        int openCount = 0;
        string out;

        for(auto &c : s){
            if(c == ')') {
                closeCount--;
                if(!openCount) continue;
                openCount--;
            }
            else if(c == '('){
                if(openCount == closeCount) continue;
                openCount++;
            }
            out.push_back(c);
        }

        return out;
    }
};

File: Algorithms/Data Structure dependent/Stacks and Queue/minRemoveToMakeValid.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        closeCount = 0

        for c in s:
            if c == ')': closeCount += 1

        openMatchedCount = 0
        out = ""

        for c in s:
            if c == ')':
                closeCount -= 1
                if not openMatchedCount: continue
                openMatchedCount -= 1
            elif c == '(':
                if openMatchedCount == closeCount: continue
                openMatchedCount += 1
            out += c 

        return out